cdap-api-spark/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContext.java

```diff
@@ -0,0 +1,208 @@
+import co.cask.cdap.api.data.batch.Split;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.flow.flowlet.StreamEvent;
+import co.cask.cdap.api.metrics.Metrics;
+import co.cask.cdap.api.plugin.PluginContext;
+import co.cask.cdap.api.workflow.Workflow;
+import co.cask.cdap.api.workflow.WorkflowToken;
+import org.apache.spark.Partition;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.PairFunction;
+import scala.Tuple2;
+
+import java.io.Serializable;
+import java.nio.charset.Charset;
+import java.util.Collections;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Spark program execution context. User Spark program can interact with CDAP through this context.
+ */
+public abstract class JavaSparkExecutionContext implements RuntimeContext {
+
+  /**
+   * @return The specification used to configure this {@link Spark} job instance.
+   */
+  public abstract SparkSpecification getSpecification();
+
+  /**
+   * Returns the logical start time of this Spark job. Logical start time is the time when this Spark
+   * job is supposed to start if this job is started by the scheduler. Otherwise it would be the current time when the
+   * job runs.
+   *
+   * @return Time in milliseconds since epoch time (00:00:00 January 1, 1970 UTC).
+   */
+  public abstract long getLogicalStartTime();
+
+  /**
+   * Returns a {@link DatasetContext} to provide access to {@link Dataset}
+   * which can be passed in Spark program's closures.
+   *
+   * @return A {@link Serializable} {@link DatasetContext}
+   */
+  public abstract DatasetContext getDatasetContext();
+
+  /**
+   * Returns a {@link Serializable} {@link ServiceDiscoverer} for Service Discovery in Spark Program which can be
+   * passed in Spark program's closures.
+   *
+   * @return A {@link Serializable} {@link ServiceDiscoverer}
+   */
+  public abstract ServiceDiscoverer getServiceDiscoverer();
+
+  /**
+   * Returns a {@link Serializable} {@link Metrics} which can be used to emit custom metrics from user's {@link Spark}
+   * program. This can also be passed in Spark program's closures and workers can emit their own metrics
+   *
+   * @return {@link Serializable} {@link Metrics} for {@link Spark} programs
+   */
+  public abstract Metrics getMetrics();
+
+  /**
+   * Returns a {@link Serializable} {@link PluginContext} which can be used to request for plugins instances. The
+   * instance returned can also be used in Spark program's closures.
+   *
+   * @return A {@link Serializable} {@link PluginContext}.
+   */
+  public abstract PluginContext getPluginContext();
+
+  /**
+   * @return the {@link WorkflowToken} associated with the current {@link Workflow},
+   * if the {@link Spark} program is executed as a part of the Workflow.
+   */
+  @Nullable
+  public abstract WorkflowToken getWorkflowToken();
+
+  /**
+   * Creates a {@link JavaPairRDD} from the given {@link Dataset}.
```

Comment:
Good question. While these methods are more or less copied from existing Spark API, in fact exception will be thrown if the dataset or stream doesn't exists. In fact, Spark, which is native in Scala, doesn't care too much about exception (and you don't see exception in Scala program that often since all exceptions are runtime exception). In fact in Spark code base, most of the time they just throw `SparkException` and that will terminate the job.

I'll update the doc to make it clear.